In [4]:
import numpy as np
import pandas as pd
import time

from functools import partial
from scipy.stats import randint, uniform
from task_2_multiarmed_bandit.sim_lib import simulation
pd.options.mode.chained_assignment = None


In [5]:
def eps_greedy(history: pd.DataFrame, eps: float):
    if uniform.rvs() < eps:
        n = history.shape[0]
        return history.index[randint.rvs(0, n)]

    ctr = history['clicks'] / (history['impressions'] + 10)
    n = np.argmax(ctr)
    return history.index[n]

policy = partial(eps_greedy, eps=0.08)

In [6]:
# seed for homework
np.random.seed(seed=384758917)

start = time.time()
output = simulation(policy, n=200000)
end = time.time()
end - start

1 impressions have been simulated
10001 impressions have been simulated
20001 impressions have been simulated
30001 impressions have been simulated
40001 impressions have been simulated
50001 impressions have been simulated
60001 impressions have been simulated
70001 impressions have been simulated
80001 impressions have been simulated
90001 impressions have been simulated
100001 impressions have been simulated
110001 impressions have been simulated
120001 impressions have been simulated
130001 impressions have been simulated
140001 impressions have been simulated
150001 impressions have been simulated
160001 impressions have been simulated
170001 impressions have been simulated
180001 impressions have been simulated
190001 impressions have been simulated


148.6971080303192

In [7]:
# baseline regret
output['regret'], output['regret']/output['rounds'],  output['total_banners']

(2792.237649427154, 0.01396118824713577, 174)

In [12]:
def upper_confidence_bound(history: pd.DataFrame, balance: float = 1.0):
    t = history['impressions'].sum()
    exploitation_part = history['clicks'] / (history['impressions'] + 1)
    exploration_part = np.sqrt(2 * np.log(t + 1) / (history['impressions'] + 1))

    n = np.argmax(exploitation_part + balance * exploration_part)
    return history.index[n]

In [13]:
np.random.seed(seed=384758917)
# balance fine tuning
for balance in [0.01, 0.1, 0.2]:
    print(f"balance = {balance}")
    policy = partial(upper_confidence_bound, balance=balance)
    output = simulation(policy, n=200000)
    print(output['regret'], output['regret']/output['rounds'],  output['total_banners'])


balance = 0.01
1 impressions have been simulated
10001 impressions have been simulated
20001 impressions have been simulated
30001 impressions have been simulated
40001 impressions have been simulated
50001 impressions have been simulated
60001 impressions have been simulated
70001 impressions have been simulated
80001 impressions have been simulated
90001 impressions have been simulated
100001 impressions have been simulated
110001 impressions have been simulated
120001 impressions have been simulated
130001 impressions have been simulated
140001 impressions have been simulated
150001 impressions have been simulated
160001 impressions have been simulated
170001 impressions have been simulated
180001 impressions have been simulated
190001 impressions have been simulated
7219.869082625891 0.03609934541312945 185
balance = 0.1
1 impressions have been simulated
10001 impressions have been simulated
20001 impressions have been simulated
30001 impressions have been simulated
40001 impressio

С balance 0.1 получается лучший результат, который бьет baseline 857 vs 2792